engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java

```diff
@@ -460,4 +265,282 @@ Type getType() {
+     */
+    public Type getAssociatedType() {
+      return associatedType;
+    }
+
+    public Object getPolyglotSymbol(String symbolName) {
+      return polyglotSymbols.get(symbolName);
+    }
+
+    /**
+     * Create a copy of this `ModuleScope` while taking into account only the provided list of
+     * types.
+     *
+     * @param typeNames list of types to copy to the new scope
+     * @return a copy of this scope modulo the requested types
+     */
+    public ModuleScope.Builder withTypes(List<String> typeNames) {
+      Map<String, Object> polyglotSymbols = new LinkedHashMap<>(this.polyglotSymbols);
+      Map<String, Type> requestedTypes = new LinkedHashMap<>(this.types);
+      Map<Type, Map<String, Supplier<Function>>> methods = new LinkedHashMap<>();
+      Map<Type, Map<Type, Function>> conversions = new LinkedHashMap<>();
+      Set<ModuleScope.Builder> imports = new LinkedHashSet<>(this.imports);
+      Set<ModuleScope.Builder> exports = new LinkedHashSet<>(this.exports);
+      this.types
+          .entrySet()
+          .forEach(
+              entry -> {
+                if (typeNames.contains(entry.getKey())) {
+                  requestedTypes.put(entry.getKey(), entry.getValue());
+                }
+              });
+      Collection<Type> validTypes = requestedTypes.values();
+      this.methods.forEach(
+          (tpe, meths) -> {
+            if (validTypes.contains(tpe)) {
+              methods.put(tpe, meths);
+            }
+          });
+      this.conversions.forEach(
+          (tpe, meths) -> {
+            if (validTypes.contains(tpe)) {
+              conversions.put(tpe, meths);
+            }
+          });
+
+      return new ModuleScope.Builder(
+          module,
+          associatedType,
+          polyglotSymbols,
+          requestedTypes,
+          methods,
+          conversions,
+          imports,
+          exports);
+    }
+
+    /**
+     * Returns an already built scope.
+     *
+     * <p>The method assumes that the builder has already been built and can be treated as immutable
+     * `ModuleScope`.
+     *
+     * @return `ModuleScope` of this builder
+     */
+    public ModuleScope built() {
+      assert moduleScope != null;
+      return moduleScope;
+    }
+
+    public ModuleScope build() {
+      if (moduleScope == null) {
+        synchronized (this) {
+          if (moduleScope == null) {
+            moduleScope =
+                new ModuleScope(
+                    module,
+                    associatedType,
+                    polyglotSymbols,
+                    types,
+                    methods,
+                    conversions,
+                    imports,
+                    exports);
+          }
+        }
+      }
+      return moduleScope;
+    }
+
+    public void reset() {
+      polyglotSymbols = new LinkedHashMap<>();
+      // can't clear types because on recompilation methods etc will be assigned to the new one
```

Comment:
As suggested elsewhere, I would use `Module` as _eventually consistent_ view of `ModuleScope`. I'd add:
```java
class Module {
  /** may return different instance of scope (when reparsing happened), but always consistenly filled from the latest successfull (re)parse.
  ModuleScope getModuleScope();

  /** Start building new {@link #getModuleScope()} for this Module. Records current module state for later
  * {@code build()} call on the builder.
  * 
 * @return new builder to create and set new ModuleScope into this Module. Reuses references to existing {@code types}.
 */
  ModuleScope.Builder newBuilder();
}

/** Creates new {@code ModuleScope} and atomically sets it to associated {@code Module}. Builder is made
* unusable after calling its {@code build()} method.
* @throws ConcurrentModificationException if the {@code Module} was modified since the creation of the builder by some other builder
*/
void ModuleScope.Builder.build();

Reply:
I recall why I couldn't do it this way - when constructing truffle nodes (`IrToTruffle`) we occasionally reference modules. Those modules are still wip, meaning that their `ModuleScope` is not available, yet, but we need it to resolve methods. I will still try to figure out if we can get something vaguely similar to the above requirement.

Reply:
> Those modules are still wip, meaning that their ModuleScope is not available

What do you mean by wip? 
- is `Module` instance available? Then keep pointer to it, it will get `ModuleScope` eventually...
- ok, `ModuleScope` isn't yet ready, but it will be eventually
- _"we need it to resolve methods"_ - how can you resolve methods without finishing resolution of `ModuleScope`?

Is the code resolving methods based on partially resolved `ModuleScope`? That seems buggy to even consider...